service/worker/archiver/activities.go

```diff
@@ -98,56 +93,75 @@ func uploadHistoryActivity(ctx context.Context, request ArchiveRequest) (err err
+	historyBlobReader := container.HistoryBlobReader
+	if historyBlobReader == nil { // only will be set by testing code
+		historyBlobReader = NewHistoryBlobReader(NewHistoryBlobIterator(request, container, domainName, clusterName))
 	}
 	blobstoreClient := container.Blobstore
-	for historyBlobItr.HasNext() {
-		historyBlob, err := nextBlob(ctx, historyBlobItr)
-		if err != nil {
-			logging.LogFailArchivalUploadAttempt(logger, err, "could not get next history blob from iterator", bucket, "")
-			return err
-		}
-		key, err := NewHistoryBlobKey(request.DomainID, request.WorkflowID, request.RunID, *historyBlob.Header.CurrentPageToken)
+	handledLastBlob := false
+	for pageToken := common.FirstBlobPageToken; !handledLastBlob; pageToken++ {
+		key, err := NewHistoryBlobKey(request.DomainID, request.WorkflowID, request.RunID, pageToken)
 		if err != nil {
 			logging.LogFailArchivalUploadAttempt(logger, err, "could not construct blob key", bucket, "")
 			return cadence.NewCustomError(errConstructBlob)
 		}
-		if exists, err := blobExists(ctx, blobstoreClient, bucket, key); err != nil {
-			logging.LogFailArchivalUploadAttempt(logger, err, "could not check if blob already exists", bucket, key.String())
+		tags, err := getTags(ctx, blobstoreClient, bucket, key)
+		if err != nil && err != blobstore.ErrBlobNotExists {
+			logging.LogFailArchivalUploadAttempt(logger, err, "could not get blob tags", bucket, key.String())
+			return err
+		}
+		runConstTest := false
+		if err == nil {
+			handledLastBlob = IsLast(tags)
+			// this is a sampling based sanity check used to ensure deterministic blob construction
+			// is operating as expected, the correctness of archival depends on this deterministic construction
+			runConstTest = runConstructionCheck(container.Config.DeterministicConstructionCheckProbability())
+			if !runConstTest {
+				continue
+			}
+		}
+		historyBlob, err := getBlob(ctx, historyBlobReader, pageToken)
+		if err != nil {
+			logging.LogFailArchivalUploadAttempt(logger, err, "could not get history blob from reader", bucket, "")
 			return err
-		} else if exists {
-			logging.LogBlobAlreadyUploaded(logger, bucket, key.String())
-			encounteredBlobAlreadyExists = true
-			continue
 		}
 		blob, reason, err := constructBlob(historyBlob, container.Config.EnableArchivalCompression(domainName))
 		if err != nil {
 			logging.LogFailArchivalUploadAttempt(logger, err, reason, bucket, key.String())
 			return cadence.NewCustomError(errConstructBlob)
 		}
+		if runConstTest {
+			existingBlob, err := downloadBlob(ctx, blobstoreClient, bucket, key)
+			if err != nil {
+				logger.WithError(err).Error("failed to download blob for deterministic construction verification")
+				metricsClient.IncCounter(metrics.ArchiverUploadHistoryActivityScope, metrics.ArchiverCouldNotRunDeterministicConstructionCheckCount)
+			} else if !blob.Equal(existingBlob) {
+				logger.Error("deterministic construction check failed")
```

Comment:
Can you make sure this logger is tagged correctly with all necessary fields like domain, workflow, run, etc?

Reply:
It is at the top of method we assign logger to result of this method

```
func tagLoggerWithRequest(logger bark.Logger, request ArchiveRequest) bark.Logger {
	return logger.WithFields(bark.Fields{
		logging.TagArchiveRequestDomainID:             request.DomainID,
		logging.TagArchiveRequestWorkflowID:           request.WorkflowID,
		logging.TagArchiveRequestRunID:                request.RunID,
		logging.TagArchiveRequestEventStoreVersion:    request.EventStoreVersion,
		logging.TagArchiveRequestBranchToken:          string(request.BranchToken),
		logging.TagArchiveRequestNextEventID:          request.NextEventID,
		logging.TagArchiveRequestCloseFailoverVersion: request.CloseFailoverVersion,
	})
}
```